os/kernel/log_dump/log_dump.c

```diff
@@ -0,0 +1,206 @@
+
+	if (node == NULL) {
+		return LOG_DUMP_MEM_FAIL;
+	}
+
+	curbytes = 0;
+	current_ptr = 0;
+	sq_addfirst((sq_entry_t *)node, &log_dump_chunks);
+	/* set the size of chunks to 1 chunk */
+	log_dump_size = LOG_CHUNK_SIZE;
+
+	return LOG_DUMP_OK;
+}
+
+void log_dump_read_init(void)
+{
+	read_node = (struct log_dump_chunk_s *)sq_peek(&log_dump_chunks);	/* reset the read node to head */
+	current_ptr = 0;							/* reset the read pointer */
+}
+
+static void log_dump_mem_check(size_t max_size)
+{
+	if (log_dump_size > max_size) {
+		/* remove extra chunks from start */
+		size_t extra_chunk_size = log_dump_size - max_size;
+		int extra_chunks_count = extra_chunk_size / LOG_CHUNK_SIZE;
+		struct log_dump_chunk_s *next_chunk;
+		
+		for (int i = 0; i < extra_chunks_count; i++) {
+			next_chunk = (struct log_dump_chunk_s *)sq_remfirst(&log_dump_chunks);
+			current_ptr = 0; /* reset the read pointer as the head is removed */
+			kmm_free(next_chunk);
+		}
+	}
+}
+
+int log_dump_save(char ch)
+{
+	/* need to check if the current chunks size is over max_log_size or greater than x% of free heap */
+	struct mm_heap_s *heap = kmm_get_heap(); /* get heap details using the head of the chunks linked list */
+
+	size_t free_size = heap->mm_heapsize - heap->total_alloc_size;                  /* get the free size */
+	size_t percent_free_size = (free_size * CONFIG_LOG_DUMP_MAX_FREE_HEAP) / 100;
+	size_t max_size = MIN(percent_free_size, CONFIG_LOG_DUMP_MAX_SIZE);             /* setting the upper limit */
+	max_size = MAX(max_size, CONFIG_LOG_DUMP_MIN_SIZE);                             /* checking the lower limit */
+
+	struct log_dump_chunk_s *log_dump_tail = (struct log_dump_chunk_s *)sq_tail(&log_dump_chunks);
+
+	if (curbytes == CONFIG_LOG_DUMP_CHUNK_SIZE) { /* last chunk is full */
+
+		if (log_dump_size + LOG_CHUNK_SIZE < max_size) {	
+			/* memory available, so fetch a new buffer */
+			struct log_dump_chunk_s *node = (struct log_dump_chunk_s *)kmm_malloc(LOG_CHUNK_SIZE);
+
+			if (node == NULL) {
+				return LOG_DUMP_MEM_FAIL;
+			}
+
+			curbytes = 1;
+			node->arr[0] = ch;
+			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			log_dump_size += LOG_CHUNK_SIZE;
+
+		} else {
+			/* logs reached memory limit, reuse the head and free extra logs */
+			struct log_dump_chunk_s *node = (struct log_dump_chunk_s *)sq_remfirst(&log_dump_chunks);
+			curbytes = 1;
+			current_ptr = 0; /* reset the read pointer as head is modified */
+			node->arr[0] = ch;
+			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			log_dump_mem_check(max_size);
+		}
+
+	} else { /* there is place in tail */
+
+		log_dump_tail->arr[curbytes] = ch;
+		curbytes += 1;
+		if (curbytes % CONFIG_LOG_DUMP_MEMCHECK_SIZE == 0) {
+			log_dump_mem_check(max_size);
+		}
+	}
+
+	return LOG_DUMP_OK;
+}
+
+size_t log_dump_read(FAR char *buffer, size_t buflen)
+{
+	size_t ret = 0;
+	irqstate_t state;
+
+	state = irqsave(); /* to ensure that the read is not disturbed by add_char */
```

Comment:
How about sched_lock and unlock instead of irqsave and restore? Because we must guarantee micom uart tx. Irqsave and restore corrupts but sched_lock and unlock doesn't.

Reply:
Done. modifed to sched_lock.